sway/input/text_input.c

```diff
@@ -255,6 +264,207 @@ static void relay_handle_text_input(struct wl_listener *listener,
 	sway_text_input_create(relay, wlr_text_input);
 }
 
+static void input_popup_update(struct sway_input_popup *popup) {
+	struct sway_text_input *text_input =
+		relay_get_focused_text_input(popup->relay);
+
+	if (text_input == NULL || text_input->input->focused_surface == NULL) {
+		return;
+	}
+
+	if (!popup->popup_surface->surface->mapped) {
+		return;
+	}
+
+	if (popup->scene_tree != NULL) {
+		wlr_scene_node_destroy(&popup->scene_tree->node);
+	}
+
+	bool cursor_rect = text_input->input->current.features
+		& WLR_TEXT_INPUT_V3_FEATURE_CURSOR_RECTANGLE;
+	struct wlr_surface *focused_surface = text_input->input->focused_surface;
+	struct wlr_box cursor = text_input->input->current.cursor_rectangle;
+
+	struct sway_output *swayoutput;
+	struct wlr_box output_box;
+	struct wlr_box parent;
+	struct wlr_layer_surface_v1 *layer_surface =
+		wlr_layer_surface_v1_try_from_wlr_surface(focused_surface);
+
+	if (layer_surface != NULL) {
+		struct sway_layer_surface *layer =
+			layer_surface->data;
+
+		struct wlr_output* output = layer->layer_surface->output;
+		swayoutput = output_from_wlr_output(output);
+		wlr_output_layout_get_box(root->output_layout, output, &output_box);
+		int lx, ly;
+		wlr_scene_node_coords(&layer->tree->node, &lx, &ly);
+		parent.x = lx - swayoutput->scene_output->x;
+		parent.y = ly - swayoutput->scene_output->y;
+
+	} else {
+		struct sway_view *view = view_from_wlr_surface(focused_surface);
+		int lx, ly;
+		wlr_scene_node_coords(&view->scene_tree->node, &lx, &ly);
+		struct wlr_output *output = wlr_output_layout_output_at(root->output_layout,
+			view->container->pending.content_x + view->geometry.x,
+			view->container->pending.content_y + view->geometry.y);
+		swayoutput = output_from_wlr_output(output);
+		wlr_output_layout_get_box(root->output_layout, output, &output_box);
+		parent.x = lx - swayoutput->scene_output->x;
+		parent.y = ly - swayoutput->scene_output->y;
+
+		parent.width = view->geometry.width;
+		parent.height = view->geometry.height;
+	}
+	popup->scene_tree = wlr_scene_subsurface_tree_create(swayoutput->layers.shell_top, popup->popup_surface->surface);
```

Comment:
We need to ensure that the scene node is only visible after the surface is mapped and ensure that it becomes disabled when it is unmapped. wlroots may destroy the surface late so that means we might show an unmapped surface. Interacting with unmapped surface often leads to crashes.

Reply:
Then should I add a new layer in output.h?

Reply:
No, `root.h`. It seems like the logic would have an easier time if things were based on global coordinates instead of output coordinates.

Reply:
but when I bind it to &root->root_scene_tree , the popup is not clickable anymore

Reply:
It should be bound to `layer_tree` instead, that's the tree that the cursor logic will query.

Reply:
another problem.. before I try to set it to root, I found that although I can click the popup, but when I focus on popup, it removed the focus on textline, then after I click the popup, the text not fill the textline, because focus is not there anymore. How can I fix it?

Reply:
> It should be bound to `layer_tree` instead, that's the tree that the cursor logic will query.

Ok, fixed

Reply:
A scene descriptor can be assigned to the popup so that the cursor logic will not remove focus from the view.

Snippet from view.c:
```
scene_descriptor_assign(&view->scene_tree->node,
			SWAY_SCENE_DESC_VIEW, view)
```

Reply:
Thanks!